include/musica/micm.hpp

```diff
@@ -162,20 +202,126 @@ namespace musica
+    using Rosenbrock = micm::Solver<RosenbrockVectorType, micm::State<DenseMatrixVector, SparseMatrixVector>>;
+    std::unique_ptr<Rosenbrock> rosenbrock_;
 
-    std::unique_ptr<Rosenbrock> solver_;
+    /// @brief Standard-ordered Rosenbrock solver type
+    using DenseMatrixStandard = micm::Matrix<double>;
+    using SparseMatrixStandard = micm::SparseMatrix<double, micm::SparseMatrixStandardOrdering>;
+    using RosenbrockStandardType = typename micm::RosenbrockSolverParameters::
+        template SolverType<micm::ProcessSet, micm::LinearSolver<SparseMatrixStandard, micm::LuDecomposition>>;
+    using RosenbrockStandard = micm::Solver<RosenbrockStandardType, micm::State<DenseMatrixStandard, SparseMatrixStandard>>;
+    std::unique_ptr<RosenbrockStandard> rosenbrock_standard_;
 
+   private:
     std::unique_ptr<micm::SolverParameters> solver_parameters_;
   };
 
+  template<class T>
+  inline void MICM::Solve(
+      std::unique_ptr<T> &solver,
+      double time_step,
+      double temperature,
+      double pressure,
+      double air_density,
+      int num_concentrations,
+      double *concentrations,
+      int num_custom_rate_parameters,
+      double *custom_rate_parameters,
+      String *solver_state,
+      SolverResultStats *solver_stats,
+      Error *error)
+  {
+    try
+    {
+      micm::State state = solver->GetState();
+
+      for (std::size_t i{}; i < MICM_NUM_GRID_CELLS; i++)
+      {
+        state.conditions_[i].temperature_ = temperature;
+        state.conditions_[i].pressure_ = pressure;
+        state.conditions_[i].air_density_ = air_density;
+      }
+
+      state.variables_.AsVector().assign(concentrations, concentrations + num_concentrations);
+      state.custom_rate_parameters_.AsVector().assign(
+          custom_rate_parameters, custom_rate_parameters + num_custom_rate_parameters);
+
+      solver->CalculateRateConstants(state);
+      auto result = solver->Solve(time_step, state);
+
+      *solver_state = CreateString(micm::SolverStateToString(result.state_).c_str());
+
+      *solver_stats = SolverResultStats(
+          result.stats_.function_calls_,
+          result.stats_.jacobian_updates_,
+          result.stats_.number_of_steps_,
+          result.stats_.accepted_,
+          result.stats_.rejected_,
+          result.stats_.decompositions_,
+          result.stats_.solves_,
+          result.stats_.singular_,
+          result.final_time_);
+
+      for (int i = 0; i < state.variables_.AsVector().size(); i++)
+      {
+        concentrations[i] = state.variables_.AsVector()[i];
+      }
+
+      DeleteError(error);
+      *error = NoError();
+    }
+    catch (const std::system_error &e)
+    {
+      DeleteError(error);
+      *error = ToError(e);
+    }
+  }
+
+  template<class T>
+  inline std::map<std::string, std::size_t> MICM::GetSpeciesOrdering(std::unique_ptr<T> &solver, Error *error)
```

Comment:
`auto &solver` here too?

Reply:
When I replaced the template parameter with `auto`, it mostly passed the tests except for the following four. (`error: symbol not found in flat namespace`)
```
Mac / c_cxx (g++-11, gcc-11)
Windows / mingw (x64) 
Mac / c_cxx (g++-12, gcc-12)
Mac / c_cxx (g++-13, gcc-13)
```
I guess for some compilers the type of `solver` is not deducible. I found this link is useful [Why can't you use auto as a template type argument](https://stackoverflow.com/questions/18701279/why-cant-you-use-auto-as-a-template-type-argument-e-g-stdarrayauto-5) but still quite a mystery to me why it works for others.

Reply:
Adding [link](https://github.com/NCAR/musica/actions/runs/9864306616/job/27238959697?pr=172) for the error messages

Reply:
that's cool, I think it's fine as-is also